cpp/path_module/algorithm/path.cpp

```diff
@@ -1,301 +1,343 @@
+      ExpandPath(path, relationship, path_size);
+    } else if (wanted_direction == RelDirection::kBoth) {
+      if (outgoing && seen.contains({type, relationship.To().Id().AsInt()})) {
+        ExpandPath(path, relationship, path_size);
+      } else {
+        seen.insert({type, relationship.From().Id().AsInt()});
+      }
     }
   }
 }
 
 /*function used for traversal and filtering*/
-void Path::PathDFS(mgp::Path path, std::unordered_set<mgp::Relationship> &relationships_set,
-                   const mgp::RecordFactory &record_factory, int64_t path_size, const int64_t min_hops,
-                   const int64_t max_hops, const LabelSets &labelSets, const LabelBoolsStatus &labelStatus,
-                   const RelationshipSets &relationshipSets, bool &any_outgoing, bool &any_incoming) {
-  DfsByDirection(path, relationships_set, record_factory, path_size, min_hops, max_hops, labelSets, labelStatus,
-                 relationshipSets, any_outgoing, any_incoming, true);
-  DfsByDirection(path, relationships_set, record_factory, path_size, min_hops, max_hops, labelSets, labelStatus,
-                 relationshipSets, any_outgoing, any_incoming, false);
+void Path::PathExpand::DFS(mgp::Path &path, int64_t path_size) {
+  const mgp::Node node{path.GetNodeAt(path_size)};
+
+  LabelBools label_bools = helper_.GetLabelBools(node);
+  if (helper_.PathSizeOk(path_size) && helper_.AreLabelsValid(label_bools)) {
+    auto record = record_factory_.NewRecord();
+    record.Insert(std::string(kResultExpand).c_str(), path);
+  }
+
+  if (!helper_.ContinueExpanding(label_bools, path_size + 1)) {
+    return;
+  }
+
+  std::set<std::pair<std::string_view, int64_t>> seen;
+  this->Expand(path, node.InRelationships(), false, path_size, seen);
+  this->Expand(path, node.OutRelationships(), true, path_size, seen);
 }
 
-void Path::StartFunction(const mgp::Node &node, const mgp::RecordFactory &record_factory, int64_t path_size,
-                         const int64_t min_hops, const int64_t max_hops, const LabelSets &labelSets,
-                         const LabelBoolsStatus &labelStatus, const RelationshipSets &relationshipSets,
-                         bool &any_outgoing, bool &any_incoming) {
+void Path::PathExpand::StartAlgorithm(const mgp::Node node) {
   mgp::Path path = mgp::Path(node);
-  std::unordered_set<mgp::Relationship> relationships_set;
-  PathDFS(path, relationships_set, record_factory, 0, min_hops, max_hops, labelSets, labelStatus, relationshipSets,
-          any_outgoing, any_incoming);
+  DFS(path, 0);
+}
+
+void Path::PathExpand::Parse(const mgp::Value &value) {
+  if (value.IsNode()) {
+    StartAlgorithm(value.ValueNode());
+  } else if (value.IsInt()) {
+    StartAlgorithm(graph_.GetNodeById(mgp::Id::FromInt(value.ValueInt())));
+  } else {
+    throw mgp::ValueException("Invalid start type. Expected Node, Int, List[Node, Int]");
+  }
 }
```

Comment:
I think this Parse is not just parsing but also doing expansions. It make sense that this just returns node, and then we startAlgorithm in Path::Exapnd

Reply:
This will actually be useful for ExpandConfig!